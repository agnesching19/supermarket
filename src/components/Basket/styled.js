import styled from 'styled-components';

export default styled.div`
  border: 1px solid lightgray;
  border-radius: 5px;
  margin-top: 50px;
  margin: 36px 24px;
  padding: 24px;
  width: 240px;
  h2 {
    font-size: 24px;
    font-weight: normal;
    margin: 0 0 12px;
    text-align: center;
  }
`;