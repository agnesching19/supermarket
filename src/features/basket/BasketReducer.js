import { ADD_ITEM, GET_TOTAL, REMOVE_ITEM } from './actionTypes';

const initialState = {
  basketItems: [],
  total: {},
  showCheckout: false,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case ADD_ITEM:
      return { 
        ...state,
        basketItems: action.basketItems,
        showCheckout: action.showCheckout,
        total: {
          ...state.total,
          totalBeforeDiscount: action.totalBeforeDiscount,
        }
      };
    case REMOVE_ITEM:
      return {
        ...state,
        basketItems: action.basketItems,
        showCheckout: action.showCheckout,
        total: {
          ...state.total,
          totalBeforeDiscount: action.totalBeforeDiscount,
        }
      };
    case GET_TOTAL:
      return {
        ...state,
        total: action.payload,
        showCheckout: action.showCheckout,
      };
    default:
      return state;
  }
};
