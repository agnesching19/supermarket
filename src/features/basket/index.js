export { default as BasketReducer } from './BasketReducer';
export { default as useActions } from './actionCreators';
export { default as useBasketValue } from './selectors';
