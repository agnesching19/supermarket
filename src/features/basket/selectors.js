import { useSelector } from 'react-redux';

const useTotalValue = () => useSelector(state => state.checkout.total);
const useShowCheckoutValue = () => useSelector(state => state.checkout.showCheckout);
const useBasketValue = () => useSelector(state => state.checkout.basketItems);

export { useBasketValue, useTotalValue, useShowCheckoutValue };
