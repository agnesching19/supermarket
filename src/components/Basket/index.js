import React, { useState, useCallback }  from 'react';
import { uniqueId } from 'lodash';
import { useActions, useTotalValue, useBasketValue, useShowCheckoutValue } from '../../features/basket';
import Styled from './styled';

const Basket = () => {
  const total = useTotalValue();
  const basketItems = useBasketValue();
  const showTotal = useShowCheckoutValue();
  const { getTotal } = useActions();
  const hasDiscount = total.discount && total.discount.length > 0;
  
  const subTotal = total.totalBeforeDiscount ? total.totalBeforeDiscount.toFixed(2) : '0.00';

  const totalSavings =
    total.discount && total.discount.length > 0 && total.discount.map(d => d.saved).reduce((ac, c) => ac + c, 0);

  const getPriceByWeightIndex = () => basketItems
    .map(i => i)
    .findIndex((f => f.priceByWeight === true));

  const hasWeightItem = basketItems[getPriceByWeightIndex()] ? basketItems[getPriceByWeightIndex()].weight > 0 : null;

  return (
    <Styled>
      <h2>Basket</h2>
      <ul className='list'>
        {basketItems.map(i => {
          if (i.priceByWeight) {
            return !!hasWeightItem && (
              <li key={uniqueId()}>
                <span className='item-total weight'>
                  <span>
                    {i.name} - <span className='byWeight'>{i.weight.toFixed(2)}kg @ £{i.unitPrice}/kg</span>
                  </span>
                  <span className='item-price'>
                    {`£${(i.weight * i.unitPrice).toFixed(2)}`}
                  </span>
                </span>
              </li>
            )}
          return (
            <li key={uniqueId()}>
              <span className='item-total'>
                <span>{i.name}</span>
                <span className='item-price'>{`£${i.unitPrice ? i.unitPrice.toFixed(2) : '0.00'}`}</span>
              </span>
            </li>  
          )
       })}
      </ul>

      <hr/ >
      <p className='sub-total'>
        <span>Sub total</span>
        <span>{`£${subTotal}`}</span>
      </p>

      <button
        type="button"
        onClick={getTotal}
        className='btn-checkout'
      >
        Checkout
      </button>

      {basketItems.length > 0 && (
        <>
          <ul className="total">
            {showTotal && (
              <>
                {hasDiscount && (
                  <>
                    <hr />
                    <li className='title'>Savings</li>
                  </>
                )}
                {total.discount.map(m => (
                  <li key={uniqueId()}>
                    <span className='item-total'>
                      <span>{m.name}</span>
                      <span className='item-price'>-£{m.saved}</span>
                    </span>
                  </li>
                ))}
                {hasDiscount && (
                  <>
                  <hr className='savings-bottom' />
                  <li>
                    <span className='item-total'>
                      <span>Total savings</span>
                      <span className='item-price'>
                        -£{totalSavings}
                      </span>
                    </span>
                  </li>
                  <hr className='total-top' />
                  </>
                )}
                <li>
                  <span className='item-total'>
                    <span>Total to Pay</span>
                    <span className='item-price'>
                      £{total.totalFinal}
                    </span>
                  </span>
                </li>
              </>
            )}
          </ul>
        </>
      )}
    </Styled>
  );
};

export default Basket;
