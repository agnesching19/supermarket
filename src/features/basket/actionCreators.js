import {useCallback} from 'react';
import {useDispatch} from 'react-redux';
import { useBasketValue } from './selectors';
import { REMOVE_ITEM, ADD_ITEM, GET_TOTAL } from './actionTypes';

const useActions = () => {
  const dispatch = useDispatch();
  const basketItems = useBasketValue();

  const getIndex = useCallback(
    id => basketItems.map(i => i).findIndex(f => f.id === id),
    [basketItems]
  );

  const subTotal = basketItems
    .map(i => {
      if (i.priceByWeight) {
        return i.unitPrice * i.weight;
      }
      return i.unitPrice;
    })
    .reduce((ac, c) => ac + c, 0);

  /** Add item action */
  const addItem = useCallback(
    (id, name, priceByItem, priceByWeight, unitPrice, promotion) => {
      const index = getIndex(id);
      /** If basket has this item and it's priceByWeight, increase the weight by 0.2kg */
      if (priceByWeight && index > -1) {
        const item = basketItems[index];
        const newItems = basketItems.filter((b, i) => i !== index);
        dispatch({
          type: ADD_ITEM,
          totalBeforeDiscount: subTotal + unitPrice * 0.2,
          showCheckout: false,
          basketItems: [
            ...newItems,
            {
              ...item,
              weight: item.weight + 0.2,
            },
          ],
        });
      } else {
        dispatch({
          type: ADD_ITEM,
          totalBeforeDiscount: priceByWeight
            ? subTotal + unitPrice * 0.2
            : subTotal + unitPrice,
          showCheckout: false,
          basketItems: [
            ...basketItems,
            {
              id,
              name,
              priceByItem,
              priceByWeight,
              unitPrice,
              promotion,
              weight: priceByWeight ? 0.2 : null,
            },
          ],
        });
      }
    },
    [basketItems, dispatch, getIndex, subTotal]
  );

  /** Remove item action */
  const removeItem = useCallback(
    (id, priceByWeight, unitPrice) => {
      const index = getIndex(id);
      const newItems = basketItems.filter((b, i) => i !== index);
      /** If basket has this item and it's priceByWeight, decrease the weight by 0.2kg */
      if (priceByWeight && index > -1) {
        const item = basketItems[index];
        const lastByWeightItem = item.weight.toFixed(2) <= 0.2;
        
        dispatch({
          type: REMOVE_ITEM,
          totalBeforeDiscount: lastByWeightItem
            ? subTotal
            : subTotal - unitPrice * 0.2,
          showCheckout: false,
          basketItems: [
            ...newItems,
            {
              ...item,
              weight: lastByWeightItem ? 0 : item.weight - 0.2,
            },
          ],
        });
      } else {
        const item = basketItems[index];
        const hasPriceByItem = item ? subTotal - item.unitPrice : 0
        const hasPriceByWeight = item ? subTotal - item.unitPrice * 0.2 : 0;
      
        dispatch({
          type: REMOVE_ITEM,
          showCheckout: false,
          totalBeforeDiscount: priceByWeight ? hasPriceByWeight : hasPriceByItem,
          basketItems: index > -1 ? [...newItems] : [...basketItems],
        });
      }
    },
    [basketItems, dispatch, getIndex, subTotal]
  );

  /** Calculate total */
  const getTotal = useCallback(() => {
    const getSavings = () => {
      const counts = basketItems
        .map(item => item)
        .reduce(
          (acc, value) => ({
            ...acc,
            [value.id]: (acc[value.id] || 0) + 1,
          }),
          {}
        );

      /**
       * If there are more items using the same type of discount, we can use an api to store
       * and calculate the discount amount; here I just hardcode item_coke and item_beans as
       * these are what we need in this task.
       */
      const cokeSaved = counts.item_coke >= 2 ? counts.item_coke / 2 : 0;
      const itemCoke = basketItems[getIndex('item_coke')];
      const discountCoke =
        cokeSaved > 0
          ? {
              name: itemCoke && itemCoke.promotion,
              saved: itemCoke && itemCoke.unitPrice * Math.floor(cokeSaved),
            }
          : null;

      const beansSaved = counts.item_beans >= 3 ? counts.item_beans / 3 : 0;
      const itemBeans = basketItems[getIndex('item_beans')];
      const discountBeans =
        beansSaved > 0
          ? {
              name: itemBeans && itemBeans.promotion,
              saved: itemBeans && itemBeans.unitPrice * Math.floor(beansSaved),
            }
          : null;

      const discountItems = [discountCoke, discountBeans].filter(i => i);

      return discountItems;
    };

    const totalSaved = getSavings()
      .map(s => s.saved)
      .reduce((ac, c) => ac + c, 0);

    dispatch({
      type: GET_TOTAL,
      payload: {
        discount: getSavings(),
        totalBeforeDiscount: subTotal,
        totalFinal: (subTotal - totalSaved).toFixed(2),
      },
      showCheckout: true,
    });
  }, [basketItems, dispatch, getIndex, subTotal]);

  return {addItem, removeItem, getTotal};
};

export default useActions;
