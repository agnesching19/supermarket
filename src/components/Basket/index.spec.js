import React from 'react';
import { Provider } from 'react-redux';
import { render, fireEvent } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { GET_TOTAL } from 'features/basket/actionTypes';
import Basket from './index';

describe('components > Basket', () => {
  /** Create mock store with the count value */
  const mockStore = configureStore([]);
  const store = mockStore({
    checkout: {
      total: {
        discount: [{
          name: 'Get 2 for 1!',
          saved: 0.7
        },{
          name: 'Get 3 for 2!',
          saved: 0.5
        }],
        totalBeforeDiscount: 3.69,
        totalFinal: "2.50"
      },
      basketItems: [
        {
          id: 'item_beans',
          name: 'Beans',
          promotion: 'Get 3 for 2!',
          unitPrice: 0.5,
          priceByItem: true
        },
        {
          id: 'item_beans',
          name: 'Beans',
          promotion: 'Get 3 for 2!',
          unitPrice: 0.5,
          priceByItem: true
        },
        {
          id: 'item_beans',
          name: 'Beans',
          promotion: 'Get 3 for 2!',
          unitPrice: 0.5,
          priceByItem: true
        },
        {
          id: 'item_coke',
          name: 'Coke',
          promotion: 'Get 2 for 1!',
          unitPrice: 0.7,
          priceByItem: true
        },
        {
          id: 'item_coke',
          name: 'Coke',
          promotion: 'Get 2 for 1!',
          unitPrice: 0.7,
          priceByItem: true
        },
        {
          id: 'item_orange',
          name: 'Oranges',
          unitPrice: 1.99,
          priceByWeight: true,
          weight: 0.4
        }
      ],
    },
  });

  jest.spyOn(store, 'dispatch');

  beforeEach(() => {
    store.dispatch.mockClear();
  });

  it('renders without crashing', () => {
    const { asFragment, getByTestId } = render(<Basket />, {
      wrapper: ({ children }) => <Provider store={store}>{children}</Provider>,
    });

    expect(asFragment()).toMatchSnapshot();

    expect(getByTestId('sub-total').textContent).toBe('Sub totalÂ£3.69');
  });

  it('dispatches an action on button click', () => {
    const { getByTestId } = render(<Basket />, {
      wrapper: ({ children }) => <Provider store={store}>{children}</Provider>,
    });
    fireEvent.click(getByTestId('checkout'));

    /** Check if store.dispatch method was run */
    expect(store.dispatch).toHaveBeenCalledTimes(1);

    /** Check if store.dispatch was run with correct action */
    expect(store.dispatch).toHaveBeenCalledWith({
      type: GET_TOTAL,
      payload: {
        discount: [{
          name: 'Get 2 for 1!',
          saved: 0.7
        },{
          name: 'Get 3 for 2!',
          saved: 0.5
        }],
        totalBeforeDiscount: 3.6960000000000006,
        totalFinal: "2.50",
      },
      showCheckout: true,
    });
  });
});
