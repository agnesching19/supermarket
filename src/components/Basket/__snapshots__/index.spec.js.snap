// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > Basket renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm cdEQCG"
  >
    <h2>
      Basket
    </h2>
    <ul
      class="list"
    >
      <li>
        <span
          class="item-total"
        >
          <span>
            Beans
          </span>
          <span
            class="item-price"
          >
            £0.50
          </span>
        </span>
      </li>
      <li>
        <span
          class="item-total"
        >
          <span>
            Beans
          </span>
          <span
            class="item-price"
          >
            £0.50
          </span>
        </span>
      </li>
      <li>
        <span
          class="item-total"
        >
          <span>
            Beans
          </span>
          <span
            class="item-price"
          >
            £0.50
          </span>
        </span>
      </li>
      <li>
        <span
          class="item-total"
        >
          <span>
            Coke
          </span>
          <span
            class="item-price"
          >
            £0.70
          </span>
        </span>
      </li>
      <li>
        <span
          class="item-total"
        >
          <span>
            Coke
          </span>
          <span
            class="item-price"
          >
            £0.70
          </span>
        </span>
      </li>
      <li>
        <span
          class="item-total weight"
        >
          <span>
            Oranges - 
            <span
              class="byWeight"
            >
              0.40kg @ £1.99/kg
            </span>
          </span>
          <span
            class="item-price"
          >
            £0.80
          </span>
        </span>
      </li>
    </ul>
    <hr />
    <p
      class="sub-total"
      data-testid="sub-total"
    >
      <span>
        Sub total
      </span>
      <span>
        £3.69
      </span>
    </p>
    <button
      class="btn-checkout"
      data-testid="checkout"
      type="button"
    >
      Checkout
    </button>
    <ul
      class="total"
    />
  </div>
</DocumentFragment>
`;
